REGRESSION (264433@main): Drawing small images becomes slower

shallawa opened the pull request:
#### efb71a57dadd74e59047c8eee2ad280b0812a16a
<pre>
REGRESSION (264433@main): Drawing small images becomes slower
<a href="https://bugs.webkit.org/show_bug.cgi?id=257358">https://bugs.webkit.org/show_bug.cgi?id=257358</a>
rdar://109810640

Reviewed by Simon Fraser.

Rearrange the logic of RenderBoxModelObject::decodingModeForImageDraw() such that
we call isVisibleInViewport() at the end of this function. But there is only one
exception to this. If the image has the attribute decoding=&quot;async&quot; specified, then
we have to make sure the image will not flicker. And to check that we have to call
isVisibleInViewport().

Also to fix a subtle one-time-flickering we should not rely on the layer repaint
count because new layers can be created when pinch zoom the image. We can rely
on a new flag called hasEverPaintedImages which is stored on the NodeRareData.
It is initialized to false and it is set to true when an image is drawn on its
RenderObject.

* LayoutTests/fast/images/async-image-composited-show.html:
* LayoutTests/fast/images/async-image-src-change.html:
* LayoutTests/fast/images/decode-decoding-change-image-src.html:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::hasEverPaintedImages const):
(WebCore::Node::setHasEverPaintedImages):
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/NodeRareData.h:
(WebCore::NodeRareData::hasEverPaintedImages const):
(WebCore::NodeRareData::setHasEverPaintedImages):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: <a href="https://commits.webkit.org/264679@main">https://commits.webkit.org/264679@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c1c4a00916f08b473f769884a219e1948c2b265e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8217 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8726 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8283 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8226 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10497 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8420 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11161 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8362 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9429 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7453 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10029 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6729 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7518 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15153 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7852 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7648 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11009 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8122 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6602 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7412 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7430 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2006 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11620 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7863 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -322,37 +322,39 @@ DecodingMode RenderBoxModelObject::decodingModeForImageDraw(const Image& image,
-    // <img decoding="sync"> forces synchronous decoding.
+    auto isFlickeringPossible = [&]() -> bool {
+        // Not first paint, so we have to avoid flickering anyway.
+        if (!element() && element()->hasEverPaintedImages())
```
The `!` seems wrong.

shallawa replied:
Yes this should be 
```
   if (!(element() && element()->hasEverPaintedImages()))
            return false;
```
Since this means the element has never been drawn so it is safe to say async image decoding will not cause flickering.

smfr commented on Source/WebCore/dom/NodeRareData.h:
```diff
@@ -305,10 +308,15 @@ class NodeRareData {
 
 private:
+    struct DataBitFields {
+        uint16_t isElementRareData : 1;
```
Should be `bool`, right? We don't want to use 16 bits for this bitfield.

shallawa replied:
Done and the bit fields structure was removed.

smfr commented on Source/WebCore/rendering/RenderBoxModelObject.cpp:
```diff
@@ -322,37 +322,39 @@ DecodingMode RenderBoxModelObject::decodingModeForImageDraw(const Image& image,
-    // <img decoding="async"> forces asynchronous decoding.
-    if (is<HTMLImageElement>(element())) {
+        // <img decoding="async"> forces asynchronous decoding but make sure either
+        // the renderer has not been painted yet or it is outside the viewport.
```
renderer -> element

shallawa replied:
Fixed.

shallawa committed:
REGRESSION (264433@main): Drawing small images becomes slower
https://bugs.webkit.org/show_bug.cgi?id=257358
rdar://109810640

Reviewed by Simon Fraser.

Rearrange the logic of RenderBoxModelObject::decodingModeForImageDraw() such that
we call isVisibleInViewport() at the end of this function. But there is only one
exception to this. If the image has the attribute decoding="async" specified, then
we have to make sure the image will not flicker. And to check that we have to call
isVisibleInViewport().

Also to fix a subtle one-time-flickering we should not rely on the layer repaint
count because new layers can be created when pinch zoom the image. We can rely
on a new flag called hasEverPaintedImages which is stored on the NodeRareData.
It is initialized to false and it is set to true when an image is drawn on its
RenderObject.

* LayoutTests/fast/images/async-image-composited-show.html:
* LayoutTests/fast/images/async-image-src-change.html:
* LayoutTests/fast/images/decode-decoding-change-image-src.html:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::hasEverPaintedImages const):
(WebCore::Node::setHasEverPaintedImages):
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/NodeRareData.h:
(WebCore::NodeRareData::hasEverPaintedImages const):
(WebCore::NodeRareData::setHasEverPaintedImages):
* Source/WebCore/rendering/BackgroundPainter.cpp:
(WebCore::BackgroundPainter::paintFillLayer):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::decodingModeForImageDraw const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: https://commits.webkit.org/264679@main
